src/message.c

```diff
@@ -2626,12 +2626,24 @@ msg_puts_printf(char_u *str, int maxlen)
 
     if (*p != NUL && !(silent_mode && p_verbose == 0))
     {
-	if (maxlen > 0 && STRLEN(p) > (size_t)maxlen)
-	    p[maxlen] = 0;
+	int len = (int)STRLEN(p);
+	if (maxlen > 0 && len > (size_t)maxlen)
+	    len = maxlen;
+	buf = alloc(len + 1);
```

Comment:
Is this no good?
```c
char_u c = 0;
if (maxlen > 0 && len > maxlen)
    len = maxlen;
c = p[len];
p[len] = 0;
...
p[len] = c;
```
At least, when `len <= maxlen` I think `alloc()` is no need.

Reply:
Or add `mch_msglen(str, n)` and `mch_errormsglen` ? cc: @brammool

Reply:
FYI, the way overwrite string might be crash with passing constant string.

Reply:
> FYI, the way overwrite string might be crash with passing constant string.
Ah, I had overlooked the situation...